File: apps/server/src/agent/main/skills/self-learning/scripts/saveExperience.py
import os
import sys
from datetime import datetime
from pathlib import Path

args = os.environ

type_val = args.get('type')
timestamp = args.get('timestamp') or datetime.now().isoformat().replace(':', '-').replace('.', '-')[:19]
summary = args.get('summary') or 'Experience'
content = args.get('content') or ''

if not type_val:
    print('ERROR: 缺少 type 参数')
    print('用法: saveExperience.py --type <success|failure|pattern> [--timestamp <timestamp>] [--summary <summary>] [--content <content>]')
    sys.exit(1)

valid_types = ['success', 'failure', 'pattern']
if type_val not in valid_types:
    print(f"ERROR: 无效的类型 '{type_val}'，必须是 {', '.join(valid_types)}")
    sys.exit(1)

script_dir = Path(__file__).parent
memory_dir = script_dir / '..' / 'memory'
experiences_dir = memory_dir / 'experiences'
target_dir = experiences_dir / f"{type_val}s"

target_dir.mkdir(parents=True, exist_ok=True)

filename = f"{timestamp}.md"
filepath = target_dir / filename

filepath.write_text(content, encoding='utf-8')

index_file = memory_dir / 'index.md'
type_title = type_val[0].upper() + type_val[1:]
index_entry = f"- [{timestamp}] [{summary}]({filename})"

if not index_file.exists() or index_file.stat().st_size == 0:
    initial_content = f"""# Experience Index

## {type_title}

{index_entry}
"""
    index_file.write_text(initial_content, encoding='utf-8')
else:
    index_content = index_file.read_text(encoding='utf-8')
    type_header = f"## {type_title}"

    if type_header not in index_content:
        index_content += f"\n\n## {type_title}\n\n{index_entry}"
    else:
        lines = index_content.split('\n')
        new_lines = []
        in_type_section = False
        entry_added = False

        for line in lines:
            if line.startswith('## '):
                if line == type_header:
                    in_type_section = True
                    new_lines.append(line)
                    if not entry_added:
                        new_lines.append(index_entry)
                        entry_added = True
                else:
                    in_type_section = False
                    new_lines.append(line)
            elif in_type_section and line.strip() == '' and not entry_added:
                new_lines.append(line)
                new_lines.append(index_entry)
                entry_added = True
            else:
                new_lines.append(line)

        index_content = '\n'.join(new_lines)

    index_file.write_text(index_content, encoding='utf-8')

print(f"SUCCESS: 经验已保存到 {filepath}")
print('SUCCESS: 索引已更新')

File: apps/server/src/agent/main/skills/self-learning/scripts/saveExperience.ts
import fs from 'fs'
import path from 'path'

declare namespace NodeJS {
  interface ProcessEnv {
    type: string
    timestamp?: string
    summary?: string
    content?: string
  }
}

const args = process.env

const type = args.type
const timestamp =
  args.timestamp || new Date().toISOString().replace(/[:.]/g, '-').slice(0, 19)
const summary = args.summary || 'Experience'
const content = args.content || ''

if (!type) {
  console.log('ERROR: 缺少 type 参数')
  console.log(
    '用法: saveExperience.js --type <success|failure|pattern> [--timestamp <timestamp>] [--summary <summary>] [--content <content>]',
  )
  process.exit(1)
}

const validTypes = ['success', 'failure', 'pattern']
if (!validTypes.includes(type)) {
  console.log(`ERROR: 无效的类型 '${type}'，必须是 ${validTypes.join(', ')}`)
  process.exit(1)
}

const scriptDir = __dirname
const memoryDir = path.join(scriptDir, '..', 'memory')
const experiencesDir = path.join(memoryDir, 'experiences')
const targetDir = path.join(experiencesDir, type + 's')

if (!fs.existsSync(targetDir)) {
  fs.mkdirSync(targetDir, { recursive: true })
}

const filename = `${timestamp}.md`
const filepath = path.join(targetDir, filename)

fs.writeFileSync(filepath, content, 'utf8')

const indexFile = path.join(memoryDir, 'index.md')
const typeTitle = type.charAt(0).toUpperCase() + type.slice(1)
const indexEntry = `- [${timestamp}] [${summary}](${filename})`

if (!fs.existsSync(indexFile) || fs.statSync(indexFile).size === 0) {
  const initialContent = `# Experience Index

## ${typeTitle}

${indexEntry}
`
  fs.writeFileSync(indexFile, initialContent, 'utf8')
} else {
  let indexContent = fs.readFileSync(indexFile, 'utf8')
  const typeHeader = `## ${typeTitle}`

  if (!indexContent.includes(typeHeader)) {
    indexContent += `\n\n## ${typeTitle}\n\n${indexEntry}`
  } else {
    const lines = indexContent.split('\n')
    let newLines = []
    let inTypeSection = false
    let entryAdded = false

    for (let i = 0; i < lines.length; i++) {
      const line = lines[i]

      if (line.startsWith('## ')) {
        if (line === typeHeader) {
          inTypeSection = true
          newLines.push(line)
          if (!entryAdded) {
            newLines.push(indexEntry)
            entryAdded = true
          }
        } else {
          inTypeSection = false
          newLines.push(line)
        }
      } else if (inTypeSection && line.trim() === '' && !entryAdded) {
        newLines.push(line)
        newLines.push(indexEntry)
        entryAdded = true
      } else {
        newLines.push(line)
      }
    }

    indexContent = newLines.join('\n')
  }

  fs.writeFileSync(indexFile, indexContent, 'utf8')
}

console.log(`SUCCESS: 经验已保存到 ${filepath}`)
console.log('SUCCESS: 索引已更新')
